compiler/rustc_hir/src/hir.rs

```diff
@@ -3845,6 +3855,7 @@ pub enum Node<'hir> {
     Infer(&'hir InferArg),
     WherePredicate(&'hir WherePredicate<'hir>),
     PreciseCapturingNonLifetimeArg(&'hir PreciseCapturingNonLifetimeArg),
+    UnsafeBinder(&'hir UnsafeBinderTy<'hir>),
```

Comment:
is this useful as its own `Node` variant?

Reply:
Yea, it's a binder so it needs to have late bound lifetimes associated with it iirc

Reply:
`WhereBoundPredicate` has a hir id and corresponding node kind for exactly this reason. 

Afaict `PolyTraitRef` stores its bound vars on the hir id of the trait ref it contains: https://github.com/rust-lang/rust/blob/5a6036a1802262f8cf02192b02026688d396f1d7/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs#L442

I don't believe there's a simple alternative for this case. At the end of the day we need a hir id to attach the late bound lifetimes, and if we add one we also need a corresponding node.

Reply:
The only seeming alternative would be to use the hir id of the type contained within the binder to store the late-bound vars, but that hir id is already being used for the late-bound vars for `fn` types, so `unsafe<'a> for<'b> fn()` would have issues.

Reply:
> Afaict `PolyTraitRef` stores its bound vars on the hir id of the trait ref it contains:

oh that's funny. I assumed it would store it on the `hir::Ty`'s id just like the function pointers. Weird that we have two different schemes for two things that are rather similarly stored.

> but that hir id is already being used for the late-bound vars for `fn` types, so `unsafe<'a> for<'b> fn()` would have issues.

That would use the hir id of the `UnsafeBinderTy::inner_ty`, not the one of the unsafe binder's `Ty` itself.

Reply:
hm, true, we could perhaps do that. lemme try.

Reply:
yep, i was being foolish and reusing the hir id of the *outer* ty is sufficient :)

code should be much cleaner now